compiler/front/nimconf.nim

```diff
@@ -17,25 +17,193 @@ import
   ],
   compiler/front/[
     commands,
-    msgs,
     options,
     scriptconfig
   ],
   compiler/ast/[
     lexer,
-    reports,
     idents,
     wordrecg,
     llstream,
+    lineinfos,
     ast
   ],
   compiler/utils/[
     pathutils,
   ]
 
+type
+  # TODO:
+  # - DONE: finish mapping the diags enum below
+  # - HERE: finish converting this file
+  # - then figure out how to make the handler pluggable
+  ConfigEventKind* = enum
+    ## events/errors arising from parsing and processing a compiler config file
+
+    # users errors begin
+    # expected token
+    cekParseExpectedX        ## expected some token
+    cekParseExpectedCloseX   ## expected closing ')', ']', etc
+    cekParseExpectedIdent    ## expected an identifier
+
+    # invalid input
+    cekInvalidDirective
+    # user errors end
+
+    # user output start
+    cekWriteConfig           ## write out the config
+    # user output end
+
+    # debug start
+    cekDebugTrace            ## config debug trace
+    cekDebugReadStart        ## start config file read
+    cekDebugReadStop         ## stop config file read
+    # debug end
+
+    # progress begin
+    cekProgressConfStart
+    # progress end
+
+  ConfigFileEvent* = object
+    case kind*: ConfigEventKind:
+      of cekParseExpectedX, cekParseExpectedCloseX, cekParseExpectedIdent,
+         cekInvalidDirective, cekWriteConfig, cekDebugTrace:
+        location*: TLineInfo         ## diagnostic location
+      of cekDebugReadStart, cekDebugReadStop, cekProgressConfStart:
+        discard
+    instLoc*: InstantiationInfo ## instantiation in lexer's source
+    msg*: string
+
 # ---------------- configuration file parser -----------------------------
 # we use Nim's lexer here to save space and work
 
+# xxx: importing `reports`, the whole need to bridge via `handleDiagReport`
+#      it's all terrible. this needs to be further broken up so the lexer just
+#      emits diagnostics and is configured as to how it should handle them as
+#      they arise wrt to aborting, etc
+from compiler/ast/report_enums import ReportKind, ReportCategory
+from compiler/ast/reports import Report, DebugReport, ExternalReport,
+                                 InternalReport, LexerReport, ParserReport,
+                                 toReportLineInfo
+from compiler/front/msgs import handleReport
+import std/options as std_options
+
+proc handleConfigEvent(
```

Comment:
The nimconf bit is pretty lightweight as it is today, but also not necessarily the most testable. Anyways, in order to break it up from reporting I need a way for it to signal the output events.

I'm thinking about doing something like this:

```
# same for loadConfigFile etc

proc readConfigFile(conf: ConfigRef, ...): bool =
  var
    L: Lexer
    tok: Token
    stream: PLLStream

  stream = llStreamOpen(filename, fmRead)
  if stream != nil:
   # move handleRead implementation into the options module (as a callback) then let it do the reporting config.handleRead(cekDebugReadStart, filename.string)

  # code mostly as today
```

Basically all this reporting stuff moves into options, then I chase down any dependency cycles and break it up. Thoughts?